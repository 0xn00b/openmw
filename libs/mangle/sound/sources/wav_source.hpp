#ifndef MANGLE_SOUND_WAV_SOURCE_H
#define MANGLE_SOUND_WAV_SOURCE_H

#include "../source.hpp"
#include <assert.h>

namespace Mangle {
namespace Sound {

/// WAV file decoder. Has no external library dependencies.
class WavSource : public SampleSource
{
  // Sound info
  uint32_t rate, channels, bits;

  // Total size (of output) and bytes left
  uint32_t total, left;

  // Offset in input of the beginning of the data block
  size_t dataOffset;

  Mangle::Stream::StreamPtr input;

  void open(Mangle::Stream::StreamPtr);

 public:
  /// Decode the given sound file
  WavSource(const std::string&);

  /// Decode from stream
  WavSource(Mangle::Stream::StreamPtr s)
  { open(s); }

  void getInfo(int32_t *rate, int32_t *channels, int32_t *bits);
  size_t read(void *data, size_t length);

  void seek(size_t);
  size_t tell() const { return total-left; }
  size_t size() const { return total; }
  bool eof() const { return left > 0; }
};

#include "loadertemplate.hpp"

/// A factory that loads WavSources from file and stream
typedef SSL_Template<WavSource,true,true> WavLoader;

}} // Namespace
#endif
