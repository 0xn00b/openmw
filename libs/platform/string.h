// Wrapper for string.h on Mac
#ifndef _STRING_WRAPPER_H
#define _STRING_WRAPPER_H

#include <string.h>
#ifdef __APPLE__
// need our own implementation of strnlen
static size_t strnlen(const char *s, size_t n)
{
  const char *p = (const char *)memchr(s, 0, n);
  return(p ? p-s : n);
}
#endif
#endif
