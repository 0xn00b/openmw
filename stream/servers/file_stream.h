#ifndef MANGLE_STREAM_FILESERVER_H
#define MANGLE_STREAM_FILESERVER_H

#include "std_stream.h"
#include <fstream>

namespace Mangle {
namespace Stream {

/** Very simple file input stream, based on std::ifstream
 */
class FileStream : public StdStream
{
  std::ifstream file;

 public:
  FileStream(const std::string &name)
    : StdStream(&file)
    {
      file.open(name, ios::binary);
    }
  ~FileStream() { file.close(); }
};

typedef boost::shared_ptr<FileStream> FileStreamPtr;

}} // namespaces
#endif
