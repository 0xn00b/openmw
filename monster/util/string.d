/*
  Monster - an advanced game scripting language
  Copyright (C) 2004, 2007, 2008  Nicolay Korslund
  Email: <korslund@gmail.com>
  WWW: http://monster.snaptoad.com/

  This file (string.d) is part of the Monster script language
  package.

  Monster is distributed as free software: you can redistribute it
  and/or modify it under the terms of the GNU General Public License
  version 3, as published by the Free Software Foundation.

  This program is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  General Public License for more details.

  You should have received a copy of the GNU General Public License
  version 3 along with this program. If not, see
  http://www.gnu.org/licenses/ .

 */

module monster.util.string;

import std.utf;
import std.string;

// These functions check whether a string begins or ends with a
// certain substring.
bool begins(char[] str, char[] start)
{
  if(str.length < start.length ||
     str[0..start.length] != start) return false;
  return true;
}

bool ends(char[] str, char[] end)
{
  if(str.length < end.length ||
     str[$-end.length..$] != end) return false;
  return true;
}

// Case insensitive versions of begins and ends
bool iBegins(char[] str, char[] start)
{
  if(str.length < start.length ||
     icmp(str[0..start.length], start) != 0) return false;
  return true;
}

bool iEnds(char[] str, char[] end)
{
  if(str.length < end.length ||
     icmp(str[$-end.length..$], end) != 0) return false;
  return true;
}

char[] nextWord(ref char[] str, char delim = ' ')
{
  int i = find(str, delim);
  char[] result;

  // No 'delim' found, return the entire string and set remainder to
  // null.
  if(i == -1)
    {
      result = str;
      str = null;
      return result;
    }

  // A separator was found. Return everything upto 'delim' (index i),
  // put the remainder of the string (not including the char at [i])
  // in str.
  result = str[0..i];
  str = str[i+1..$];
  return result;
}

// Strip trailing zeros
char[] stripz(char [] s)
{
  foreach(int i, char c; s)
    if( c == 0 )
      return s[0..i];

  return s;
}

// Convert a long integer into a string using nice comma
// formatting. delim is the delimiter character, size is the number of
// digits in each group. See the unittest for examples.
char[] comma(long i, char delim=',', int size = 3)
{
  char[] str = toString(i);
  char[] res;

  if(i<0) str=str[1..$];

  str.reverse;
  foreach(int j, char c; str)
    { 
      if(j!=0 && j%size == 0)
        res = delim ~ res;
      res = c ~ res;
    }

  if(i<0) res = "-" ~ res;

  return res;
}
