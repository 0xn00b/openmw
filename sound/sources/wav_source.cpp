#include "wav_source.hpp"

#include "../../tools/str_exception.hpp"
#include "../../stream/servers/file_stream.hpp"

using namespace Mangle::Stream;
using namespace Mangle::Sound;

static void fail(const std::string &msg)
{ throw str_exception("Mangle::Wav exception: " + msg); }

void WavSource::getInfo(int32_t *pRate, int32_t *pChannels, int32_t *pBits)
{
  // Use the values we found in the constructor
  *pRate = rate;
  *pChannels = channels;
  *pBits = bits;
}

void WavSource::seek(size_t pos)
{
  // Seek the stream and set 'left'
  assert(isSeekable);
  if(pos > total) pos = total;
  input->seek(dataOffset + pos);
  left = total-pos;
}

size_t WavSource::read(void *data, size_t length)
{
  if(length > left)
    length = left;
  input->read(data, length);
  return length;
}

void WavSource::open(Mangle::Stream::StreamPtr data)
{
  input = data;

  hasPosition = true;
  hasSize = true;
  // If we can check position and seek in the input stream, then we
  // can seek the wav data too.
  isSeekable = input->isSeekable && input->hasPosition;

  // Read header
  unsigned int val;

  input->read(&val,4);  // header
  if(val != 0x46464952) // "RIFF"
    fail("Not a WAV file");

  input->read(&val,4);  // size (ignored)
  input->read(&val,4);  // file format
  if(val != 0x45564157) // "WAVE"
    fail("Not a valid WAV file");

  input->read(&val,4);  // "fmt "
  input->read(&val,4);  // chunk size (must be 16)
  if(val != 16)
    fail("Unsupported WAV format");

  input->read(&val,2);
  if(val != 1)
    fail("Non-PCM (compressed) WAV files not supported");

  // Sound data specification
  channels = 0;
  input->read(&channels,2);
  input->read(&rate, 4);

  // Skip next 6 bytes
  input->read(&val, 4);
  input->read(&val, 2);

  // Bits per sample
  bits = 0;
  input->read(&bits,2);

  input->read(&val,4);  // Data header
  if(val != 0x61746164) // "data"
    fail("Expected data block");

  // Finally, read the data size
  input->read(&total,4);
  left = total;

  // Store the beginning of the data block for later
  if(input->hasPosition)
    dataOffset = input->tell();
}

WavSource::WavSource(const std::string &file)
{ open(StreamPtr(new FileStream(file))); }
