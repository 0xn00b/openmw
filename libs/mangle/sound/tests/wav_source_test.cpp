#include <iostream>

#include "../sources/wav_source.hpp"
#include "../../stream/servers/file_stream.hpp"

#include <assert.h>
#include <string.h>

using namespace std;
using namespace Mangle::Sound;
using namespace Mangle::Stream;

int main()
{
  WavSource wav("cow.wav");

  cout << "Source size: " << wav.size() << endl;
  int rate, channels, bits;
  wav.getInfo(&rate, &channels, &bits);
  cout << "rate=" << rate << "\nchannels=" << channels
       << "\nbits=" << bits << endl;

  cout << "Reading entire buffer into memory\n";
  void *buf = malloc(wav.size());
  wav.read(buf, wav.size());

  cout << "\nReading cow.raw\n";
  FileStream tmp("cow.raw");
  cout << "Size: " << tmp.size() << endl;
  void *buf2 = malloc(tmp.size());
  tmp.read(buf2, tmp.size());

  cout << "\nComparing...\n";
  if(tmp.size() != wav.size())
    {
      cout << "SIZE MISMATCH!\n";
      assert(0);
    }

  if(memcmp(buf, buf2, wav.size()) != 0)
    {
      cout << "CONTENT MISMATCH!\n";
      assert(0);
    }

  cout << "\nDone\n";
  return 0;
}
