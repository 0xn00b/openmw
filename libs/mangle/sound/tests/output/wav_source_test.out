Source size: 37502
rate=11025
channels=1
bits=16
Reading entire buffer into memory

Reading cow.raw
Size: 37502

Comparing...

Done
